pkg/controller/sesv2/tags.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/sesv2/v1alpha1"
+	awsclient "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+const (
+	errListTagsForResource = "cannot list tags"
+	errRemoveTags          = "cannot remove tags"
+	errCreateTags          = "cannot create tags"
+)
+
+// AreTagsUpToDate for spec and resourceName
+func AreTagsUpToDate(spec []*svcapitypes.Tag, current []*svcsdk.Tag) (bool, error) {
+	add, remove := DiffTags(spec, current)
+
+	return len(add) == 0 && len(remove) == 0, nil
+}
+
+// UpdateTagsForResource with resourceName
+func UpdateTagsForResource(client sesv2iface.SESV2API, spec []*svcapitypes.Tag, resourceArn *string) error {
+	current, err := ListTagsForResource(client, resourceArn)
+	if err != nil {
+		return err
+	}
+
+	add, remove := DiffTags(spec, current)
+	if len(remove) != 0 {
+		if _, err := client.UntagResource(&svcsdk.UntagResourceInput{
+			ResourceArn: resourceArn,
+			TagKeys:     remove,
+		}); err != nil {
+			return errors.Wrap(err, errRemoveTags)
+		}
+	}
+	if len(add) != 0 {
+		if _, err := client.TagResource(&svcsdk.TagResourceInput{
+			ResourceArn: resourceArn,
+			Tags:        add,
+		}); err != nil {
+			return errors.Wrap(err, errCreateTags)
+		}
+	}
+
+	return nil
+}
+
+// ListTagsForResource for the given resource
+func ListTagsForResource(client sesv2iface.SESV2API, resourceArn *string) ([]*svcsdk.Tag, error) {
+	req := &svcsdk.ListTagsForResourceInput{
+		ResourceArn: resourceArn,
+	}
+
+	resp, err := client.ListTagsForResource(req)
+	if err != nil {
+		return nil, errors.Wrap(err, errListTagsForResource)
+	}
+
+	return resp.Tags, nil
+}
+
+// DiffTags between spec and current
+func DiffTags(spec []*svcapitypes.Tag, current []*svcsdk.Tag) (addTags []*svcsdk.Tag, removeTags []*string) {
+	currentMap := make(map[string]string, len(current))
+	for _, t := range current {
+		currentMap[awsclient.StringValue(t.Key)] = awsclient.StringValue(t.Value)
+	}
+
+	specMap := make(map[string]string, len(spec))
+	for _, t := range spec {
+		key := awsclient.StringValue(t.Key)
+		val := awsclient.StringValue(t.Value)
+		specMap[key] = awsclient.StringValue(t.Value)
+
+		if currentVal, exists := currentMap[key]; exists {
+			if currentVal != val {
+				removeTags = append(removeTags, t.Key)
+				addTags = append(addTags, &svcsdk.Tag{
+					Key:   awsclient.String(key),
+					Value: awsclient.String(val),
+				})
+			}
+		} else {
+			addTags = append(addTags, &svcsdk.Tag{
+				Key:   awsclient.String(key),
+				Value: awsclient.String(val),
+			})
+		}
+	}
+
+	for _, t := range current {
+		key := awsclient.StringValue(t.Key)
+		if _, exists := specMap[key]; !exists {
+			removeTags = append(removeTags, awsclient.String(key))
+		}
+	}
+
+	return addTags, removeTags
+}
+
+// AddExternalTags to spec if they don't exist
+func AddExternalTags(mg resource.Managed, spec []*svcapitypes.Tag) []*svcapitypes.Tag {
```

Comment:
Would it be possible to have an implementation similar to this one that covers all cases including composition and ordering? https://github.com/crossplane/provider-aws/blob/8be1568/pkg/controller/ecr/repository/controller.go#L246